### PR TITLE
feat: Add `test-connection` command

### PR DESCRIPTION
**Pull request by erezrokah**
#### Summary

Work in progress for https://github.com/cloudquery/cloudquery/issues/15704 and also https://github.com/cloudquery/cloudquery-issues/issues/1170 (internal issue).

This command uses the built in `init` gRPC call to validate plugin initialization. This relies on plugins to do all the relevant checks during init. Some plugins do it, some don't so we'll need to add plugin's support for it.

Example for AWS->Postgres without AWS creds and wrong Postgres password:

```
Loading spec(s) from examples/aws-postgres.yml
Error: failed to init source aws (cloudquery/aws@v23.1.0): rpc error: code = Internal desc = failed to init plugin: failed to initialize client: no AWS accounts were successfully configured. See warning messages in the logs for details
failed to init destination postgresql (cloudquery/postgresql@v7.2.0): rpc error: code = Internal desc = failed to init plugin: failed to initialize client: failed to get current database: failed to connect to `host=localhost user=postgres database=postgres`: failed SASL auth (FATAL: password authentication failed for user "postgres" (SQLSTATE 28P01))
exit status 1
```

### CODE REVIEW

File: cli/cmd/validate.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+func NewCmdValidate() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "validate [files or directories]",
+		Short:   validateShort,
+		Long:    validateShort,
+		Example: validateExample,
+		Args:    cobra.MinimumNArgs(1),
+		RunE:    validate,
+		Hidden:  true,
```
Comment (erezrokah) [inline cli/cmd/validate.go:33]: Hidden on purpose for now to test this a bit before making it public for everyone

File: cli/cmd/validate.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+func validate(cmd *cobra.Command, args []string) error {
+	cqDir, err := cmd.Flags().GetString("cq-dir")
+	if err != nil {
+		return err
+	}
+
+	ctx := cmd.Context()
+	log.Info().Strs("args", args).Msg("Loading spec(s)")
+	fmt.Printf("Loading spec(s) from %s\n", strings.Join(args, ", "))
```
Comment (mnorbury) [inline cli/cmd/validate.go:46]: Should this be here as you have the log line before?

File: cli/cmd/validate.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+		destinationRegInferred[i] = destination.RegistryInferred()
+	}
+
+	sourceClients, err := managedplugin.NewClients(ctx, managedplugin.PluginSource, sourcePluginConfigs, opts...)
+	if err != nil {
+		return enrichClientError(sourceClients, sourceRegInferred, err)
+	}
+	defer func() {
+		if err := sourceClients.Terminate(); err != nil {
+			fmt.Println(err)
```
Comment (mnorbury) [inline cli/cmd/validate.go:105]: Can we use the logger?

**Review by mnorbury — COMMENTED** [review]
Looks good, left a few comments about potential left over `fmt` and using the logger for error messages.

File: cli/cmd/validate.go (hunk 2 above)
Comment (erezrokah) [reply cli/cmd/validate.go]: So I think we need both, see https://github.com/cloudquery/cloudquery/blob/c6e06aa2fb25010e0266538200dcb7212aa17248/cli/cmd/sync.go#L99

one goes to the log, the other shows up in the console via `stdout`

File: cli/cmd/validate.go (hunk 3 above)
Comment (erezrokah) [reply cli/cmd/validate.go]: This is similar to https://github.com/cloudquery/cloudquery/blob/c6e06aa2fb25010e0266538200dcb7212aa17248/cli/cmd/sync.go#L163, where we want to show the error in `stdout`

**Review by disq — COMMENTED** [review]
LGTM but the title should be "feat: Add `test-connection` command"?

Validate would be the offline (no-conn) version of this.

File: cli/cmd/test_connection.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	"github.com/cloudquery/cloudquery/cli/internal/auth"
+	"github.com/cloudquery/cloudquery/cli/internal/specs/v0"
+	"github.com/cloudquery/plugin-pb-go/managedplugin"
+	"github.com/cloudquery/plugin-pb-go/pb/plugin/v3"
+	"github.com/rs/zerolog/log"
+	"github.com/spf13/cobra"
+)
+
+const (
+	testConnectionShort   = "Test plugins connection to sources and destinations"
```
Comment (disq) [inline cli/cmd/test_connection.go:17]:
```suggestion
	testConnectionShort   = "Test plugin connections to sources and destinations"
```

File: cli/cmd/test_connection.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	testConnectionExample = `# Test plugins connection to sources and destinations
```
Comment (disq) [inline cli/cmd/test_connection.go:18]:
```suggestion
	testConnectionExample = `# Test plugin connections to sources and destinations
```

File: cli/cmd/test_connection.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+cloudquery test-connection ./directory
+# Test plugins connection from directories and files
```
Comment (disq) [inline cli/cmd/test_connection.go:20]:
```suggestion
# Test plugin connections from directories and files
```

**Review by disq — CHANGES_REQUESTED** [review]
Some wording changes to avoid having to use `plugin's`.

**Comment by erezrokah** [comment]
> LGTM but the title should be "feat: Add `test-connection` command"?

Ah yeah the command was renamed

**Review by disq — APPROVED** [review]

**Pull request merged**